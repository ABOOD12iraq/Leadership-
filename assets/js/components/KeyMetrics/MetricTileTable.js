/**
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import { get } from 'lodash';

/**
 * Internal dependencies
 */
import PreviewBlock from '../PreviewBlock';

export default function MetricTileTable( props ) {
	const { Widget, loading, title, subText, rows, columns, limit } = props;

	if ( loading ) {
		return (
			<Widget noPadding>
				<PreviewBlock width="100%" height="142px" padding />
			</Widget>
		);
	}

	return (
		<Widget noPadding>
			<div className="googlesitekit-km-widget-tile">
				<h3 className="googlesitekit-km-widget-tile__title">
					{ title }
				</h3>
				<div className="googlesitekit-km-widget-tile__body">
					<table className="googlesitekit-km-widget-tile__table">
						<tbody>
							{ rows
								.slice( 0, limit || rows.length )
								.map( ( row, rowIndex ) => (
									<tr key={ rowIndex }>
										{ columns.map(
											(
												{
													Component,
													field,
													className: columnClassName,
												},
												colIndex
											) => {
												const fieldValue =
													field !== undefined
														? get( row, field )
														: undefined;

												return (
													<td
														key={ colIndex }
														className={
															columnClassName
														}
													>
														{ Component && (
															<Component
																row={ row }
																fieldValue={
																	fieldValue
																}
															/>
														) }
														{ ! Component &&
															fieldValue }
													</td>
												);
											}
										) }
									</tr>
								) ) }
						</tbody>
					</table>
					{ subText && (
						<p className="googlesitekit-km-widget-tile__subtext">
							{ subText }
						</p>
					) }
				</div>
			</div>
		</Widget>
	);
}

MetricTileTable.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	loading: PropTypes.bool,
	title: PropTypes.string,
	subtext: PropTypes.string,
	rows: PropTypes.arrayOf(
		PropTypes.oneOfType( [ PropTypes.array, PropTypes.object ] )
	).isRequired,
	columns: PropTypes.arrayOf( PropTypes.object ).isRequired,
	limit: PropTypes.number,
};
