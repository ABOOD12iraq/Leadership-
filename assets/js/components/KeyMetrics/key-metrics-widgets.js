/**
 * Key Metrics widgets metadata.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import {
	KM_ANALYTICS_LOYAL_VISITORS,
	KM_ANALYTICS_NEW_VISITORS,
	KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
	KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
	KM_ANALYTICS_POPULAR_CONTENT,
	KM_ANALYTICS_POPULAR_PRODUCTS,
	KM_ANALYTICS_TOP_CITIES,
	KM_ANALYTICS_TOP_COUNTRIES,
	KM_ANALYTICS_TOP_CONVERTING_TRAFFIC_SOURCE,
	KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
} from '../../googlesitekit/datastore/user/constants';

export const KEY_METRICS_WIDGETS = {
	[ KM_ANALYTICS_LOYAL_VISITORS ]: {
		title: __( 'Loyal visitors', 'google-site-kit' ),
		description: __(
			'Portion of people who visited your site more than once',
			'google-site-kit'
		),
	},
	[ KM_ANALYTICS_NEW_VISITORS ]: {
		title: __( 'New visitors', 'google-site-kit' ),
		description: __(
			'How many new visitors you got and how the overall audience changed',
			'google-site-kit'
		),
	},
	[ KM_ANALYTICS_TOP_TRAFFIC_SOURCE ]: {
		title: __( 'Top traffic source', 'google-site-kit' ),
		description: __(
			'Channel which brought in the most visitors to your site',
			'google-site-kit'
		),
	},
	[ KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE ]: {
		title: __( 'Most engaged traffic source', 'google-site-kit' ),
		description: __(
			'Visitors coming via this channel spent the most time on your site',
			'google-site-kit'
		),
	},
	[ KM_ANALYTICS_POPULAR_CONTENT ]: {
		title: __( 'Most popular content by pageviews', 'google-site-kit' ),
		description: __(
			'Pages that brought in the most visitors',
			'google-site-kit'
		),
	},
	[ KM_ANALYTICS_POPULAR_PRODUCTS ]: {
		title: __( 'Most popular products by pageviews', 'google-site-kit' ),
		description: __(
			'Products that brought in the most visitors',
			'google-site-kit'
		),
	},
	[ KM_SEARCH_CONSOLE_POPULAR_KEYWORDS ]: {
		title: __( 'How people find your site', 'google-site-kit' ),
		description: __(
			'What people searched for before they came to your site',
			'google-site-kit'
		),
	},
	[ KM_ANALYTICS_TOP_CITIES ]: {
		title: __( 'Top cities driving traffic', 'google-site-kit' ),
		description: __(
			'Which cities you get the most visitors from',
			'google-site-kit'
		),
	},
	[ KM_ANALYTICS_TOP_COUNTRIES ]: {
		title: __( 'Top countries driving traffic', 'google-site-kit' ),
		description: __(
			'Which countries you get the most visitors from',
			'google-site-kit'
		),
	},
	[ KM_ANALYTICS_TOP_CONVERTING_TRAFFIC_SOURCE ]: {
		title: __( 'Top converting traffic source', 'google-site-kit' ),
		description: __(
			'Channel which brought in the most visits that resulted in conversions',
			'google-site-kit'
		),
	},
};
