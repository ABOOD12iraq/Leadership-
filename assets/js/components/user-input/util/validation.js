/**
 * User Input Validation Utilities.
 *
 * Site Kit by Google, Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

function trim( value ) {
	return value.trim();
}

function isEmpty( value ) {
	return value.length === 0;
}

function notEmpty( value ) {
	return value.length > 0;
}

/**
 * Returns a list of trimmed, non-empty values from the passed string array.
 *
 * @since n.e.x.t
 *
 * @param {string[]} values Array of values to trim and filter.
 * @return {string[]} Array of trimmed, non-empty values.
 */
export function getValidValues( values ) {
	return values.map( trim ).filter( notEmpty );
}

/**
 * Returns an error message for the given answer, or null if the answer is valid.
 *
 * @since n.e.x.t
 *
 * @param {string[]} values Array of values to validate.
 * @param {number}   [max]  Maximum number of values allowed. Defaults to 1.
 * @return {string|null} Error message, or null if the answer is valid.
 */
export function getErrorMessageForAnswer( values, max = 1 ) {
	if ( values.length === 0 ) {
		return max === 1
			? __( 'Please select an answer', 'google-site-kit' )
			: __( 'Please select at least 1 answer', 'google-site-kit' );
	}

	const hasEmptyValue = values.map( trim ).some( isEmpty );

	if ( values.length > 0 && hasEmptyValue ) {
		return __( 'Please select a valid answer', 'google-site-kit' );
	}

	return null;
}
