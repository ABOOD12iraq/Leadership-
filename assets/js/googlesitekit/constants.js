/**
 * Core constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const SCREEN_CONTEXT_ACTIVATION = 'activation';
export const SCREEN_CONTEXT_ADMIN_BAR = 'adminBar';
export const SCREEN_CONTEXT_DASHBOARD = 'dashboard';
export const SCREEN_CONTEXT_PAGE_DASHBOARD = 'pageDashboard';
export const SCREEN_CONTEXT_MODULE_DASHBOARD = 'moduleDashboard';
export const SCREEN_CONTEXT_SETTINGS = 'settings';
export const SCREEN_CONTEXT_SPLASH = 'splash';
export const SCREEN_CONTEXT_USER_INPUT = 'userInput';
export const SCREEN_CONTEXT_WP_DASHBOARD = 'wpDashboard';
