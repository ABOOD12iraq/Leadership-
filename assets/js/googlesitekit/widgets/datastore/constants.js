/**
 * `core/widgets` data store: constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const WIDGET_AREA_STYLES = { BOXES: 'boxes', COMPOSITE: 'composite' };
export const WIDGET_WIDTHS = { QUARTER: 'quarter', HALF: 'half', FULL: 'full' };

export const CORE_WIDGETS = 'core/widgets';

// Key Metrics Widgets
export const KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE =
	'kmAnalyticsEngagedTrafficSource';
export const KM_ANALYTICS_LOYAL_VISITORS = 'kmAnalyticsLoyalVisitors';
export const KM_ANALYTICS_NEW_VISITORS = 'kmAnalyticsNewVisitors';
export const KM_ANALYTICS_POPULAR_CONTENT = 'kmAnalyticsPopularContent';
export const KM_ANALYTICS_POPULAR_PRODUCTS = 'kmAnalyticsPopularProducts';
export const KM_ANALYTICS_TOP_CITIES = 'kmAnalyticsTopCities';
export const KM_ANALYTICS_TOP_CONVERTING_TRAFFIC_SOURCE =
	'kmTopConvertingTrafficSource';
export const KM_ANALYTICS_TOP_COUNTRIES = 'kmAnalyticsTopCountries';
export const KM_ANALYTICS_TOP_TRAFFIC_SOURCE = 'kmAnalyticsTopTrafficSource';
export const KM_SEARCH_CONSOLE_POPULAR_KEYWORDS =
	'kmSearchConsolePopularKeywords';
