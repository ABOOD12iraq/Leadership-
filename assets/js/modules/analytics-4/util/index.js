/**
 * Analytics utility functions.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

export * from './is-zero-report';
// export * from './validation';
// export * from './time-column-format';
// export * from './property';

/**
 * Extracts data required for a pie chart from the Analytics 4 report information.
 *
 * @since n.e.x.t
 *
 * @param {Array}    report                    The report data.
 * @param {Object}   [options]                 Optional. Data extraction options.
 * @param {number}   [options.keyColumnIndex]  Optional. The number of a column to extract metrics data from.
 * @param {number}   [options.maxSlices]       Optional. Limit the number of slices to display.
 * @param {boolean}  [options.withOthers]      Optional. Whether to add "Others" record to the data map. Only relevant
 *                                             if `maxSlices` is passed. If passed, the final slice will be the
 *                                             "Others" slice, i.e. the number of actual row slices will be
 *                                             `maxSlices - 1`.
 * @param {Function} [options.tooltipCallback] Optional. A callback function for tooltip column values.
 * @return {Array} Extracted data.
 */
export function extractAnalyticsDataForPieChart( report, options = {} ) {
	const {
		keyColumnIndex = 0,
		maxSlices,
		withOthers = false,
		tooltipCallback,
	} = options;

	const { rows = [], totals = [] } = report || {};

	const withTooltips = typeof tooltipCallback === 'function';
	const columns = [ 'Source', 'Percent' ];
	if ( withTooltips ) {
		columns.push( {
			type: 'string',
			role: 'tooltip',
			p: {
				html: true,
			},
		} );
	}

	const totalUsers =
		totals?.[ 0 ]?.metricValues?.[ keyColumnIndex ]?.value || 0;
	const dataMap = [ columns ];

	const currentDateRangeRows = rows.filter(
		( { dimensionValues } ) => dimensionValues[ 1 ].value === 'date_range_0'
	);

	let hasOthers = withOthers;
	let rowsNumber = currentDateRangeRows.length;
	let others = 1;
	if ( maxSlices > 0 ) {
		hasOthers = withOthers && currentDateRangeRows.length > maxSlices;
		rowsNumber = Math.min(
			currentDateRangeRows.length,
			hasOthers ? maxSlices - 1 : maxSlices
		);
	} else {
		hasOthers = false;
		rowsNumber = currentDateRangeRows.length;
	}

	for ( let i = 0; i < rowsNumber; i++ ) {
		const row = currentDateRangeRows[ i ];
		const users = row.metricValues[ keyColumnIndex ].value;
		const percent = totalUsers > 0 ? users / totalUsers : 0;

		others -= percent;

		const rowData = [ row.dimensionValues[ 0 ].value, percent ];
		if ( withTooltips ) {
			const previousDateRangeRow = rows.find(
				( { dimensionValues } ) =>
					dimensionValues[ 1 ].value === 'date_range_1' &&
					dimensionValues[ 0 ].value ===
						row.dimensionValues[ 0 ].value
			);

			rowData.push(
				tooltipCallback( row, previousDateRangeRow, rowData )
			);
		}

		dataMap.push( rowData );
	}

	if ( hasOthers && others > 0 ) {
		const rowData = [ __( 'Others', 'google-site-kit' ), others ];
		if ( withTooltips ) {
			rowData.push( tooltipCallback( null, null, rowData ) );
		}

		dataMap.push( rowData );
	}

	return dataMap;
}
