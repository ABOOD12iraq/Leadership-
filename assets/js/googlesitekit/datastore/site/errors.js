/**
 * `core/site` data store: Error info.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import Data from 'googlesitekit-data';
import { STORE_NAME } from './constants';
import { createFetchStore } from '../../data/create-fetch-store';

const fetchErrorStore = createFetchStore( {
	baseName: 'error',
	controlCallback: () => {
		return API.get( 'core', 'site', 'error', undefined, {
			useCache: false,
		} );
	},
	reducerCallback: ( state, error ) => {
		return {
			...state,
			error,
		};
	},
} );

// Actions
const SET_SERVER_ERROR = 'SET_SERVER_ERROR';
const CLEAR_SERVER_ERROR = 'CLEAR_SERVER_ERROR';

const baseInitialState = {
	error: undefined,
};

const baseActions = {
	/**
	 * Sets the error.
	 *
	 * @since n.e.x.t
	 *
	 * @param {Object} error Error object.
	 * @return {Object} Redux-style action.
	 */
	setInternalServerError( error ) {
		return {
			type: SET_SERVER_ERROR,
			payload: { error },
		};
	},

	/**
	 * Clears the authentication error, if one was previously set.
	 *
	 * @since n.e.x.t
	 *
	 * @return {Object} Redux-style action.
	 */
	clearInternalServerError() {
		return {
			type: CLEAR_SERVER_ERROR,
			payload: {},
		};
	},
};

export const baseReducer = ( state, { type, payload } ) => {
	switch ( type ) {
		case SET_SERVER_ERROR: {
			return {
				...state,
				error: payload.error,
			};
		}

		case CLEAR_SERVER_ERROR: {
			return {
				...state,
				error: null,
			};
		}

		default: {
			return state;
		}
	}
};

const baseResolvers = {
	*getError() {
		const { select } = yield Data.commonActions.getRegistry();

		if ( ! select( STORE_NAME ).getError() ) {
			yield fetchErrorStore.actions.fetchGetInternalServerError();
		}
	},
};

const baseSelectors = {
	/**
	 * Gets the internal server error.
	 *
	 * @since n.e.x.t
	 * @private
	 *
	 * @param {Object} state Data store's state.
	 * @return {(Object|undefined)} Error info.
	 */
	getInternalServerError( state ) {
		const { error } = state;
		return error;
	},
};

const store = Data.combineStores(
	fetchErrorStore,
	{
		initialState: baseInitialState,
		actions: baseActions,
		reducer: baseReducer,
		resolvers: baseResolvers,
		selectors: baseSelectors,
	}
);

export const initialState = store.initialState;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

export default store;
