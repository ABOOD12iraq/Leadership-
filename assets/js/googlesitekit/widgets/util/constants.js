/**
 * Widgets layout constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { WIDGET_WIDTHS } from '../datastore/constants';

export const WIDTH_GRID_COUNTER_MAP = {
	[ WIDGET_WIDTHS.QUARTER ]: 3,
	[ WIDGET_WIDTHS.HALF ]: 6,
	[ WIDGET_WIDTHS.FULL ]: 12,
};

export const WIDTH_GRID_CLASS_MAP = {
	[ WIDGET_WIDTHS.QUARTER ]: [
		'mdc-layout-grid__cell',
		'mdc-layout-grid__cell--span-2-phone',
		'mdc-layout-grid__cell--span-3-desktop',
		'mdc-layout-grid__cell--span-4-tablet',
	],
	[ WIDGET_WIDTHS.HALF ]: [
		'mdc-layout-grid__cell',
		'mdc-layout-grid__cell--span-6-desktop',
		'mdc-layout-grid__cell--span-8-tablet',
	],
	[ WIDGET_WIDTHS.FULL ]: [
		'mdc-layout-grid__cell',
		'mdc-layout-grid__cell--span-12',
	],
};

export const HIDDEN_CLASS = 'googlesitekit-hidden';
