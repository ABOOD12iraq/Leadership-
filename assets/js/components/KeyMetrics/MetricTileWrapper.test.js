/**
 * Key Metrics MetricTileWrapper component tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
	provideUserAuthentication,
	provideUserCapabilities,
	render,
} from '../../../../tests/js/test-utils';
import { withWidgetComponentProps } from '../../googlesitekit/widgets/util';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '../../util/errors';
import { KEY_METRICS_WIDGETS } from './key-metrics-widgets';
import { KM_ANALYTICS_TOP_CATEGORIES } from '../../googlesitekit/datastore/user/constants';
import { MODULES_ANALYTICS_4 } from '../../modules/analytics-4/datastore/constants';
import MetricTileTable from './MetricTileTable';

describe( 'MetricTileWrapper', () => {
	const WidgetWithComponentProps = withWidgetComponentProps(
		KM_ANALYTICS_TOP_CATEGORIES
	)( MetricTileTable );

	it( 'renders appropriate error if required custom dimensions are not available', () => {
		const { container } = render(
			<WidgetWithComponentProps moduleSlug="analytics-4" />,
			{
				setupRegistry: ( registry ) => {
					provideUserAuthentication( registry );
					provideUserCapabilities( registry );
					registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
						propertyID: '123456789',
						availableCustomDimensions: [],
					} );
				},
				features: [ 'newsKeyMetrics' ],
			}
		);

		expect( container ).toMatchSnapshot();

		expect( container ).toHaveTextContent(
			'Update Analytics to track metric'
		);
	} );

	it( 'renders appropriate error if creating custom dimensions failed due to insufficient permissions', () => {
		const { container } = render(
			<WidgetWithComponentProps moduleSlug="analytics-4" />,
			{
				setupRegistry: ( registry ) => {
					provideUserAuthentication( registry );
					provideUserCapabilities( registry );

					registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
						propertyID: '123456789',
						availableCustomDimensions: [],
					} );

					const error = {
						code: 'test-error-code',
						message: 'Test error message',
						data: {
							reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS,
						},
					};

					registry
						.dispatch( MODULES_ANALYTICS_4 )
						.receiveCreateCustomDimensionError(
							error,
							KEY_METRICS_WIDGETS[ KM_ANALYTICS_TOP_CATEGORIES ]
								.requiredCustomDimensions?.[ 0 ]
						);
				},
				features: [ 'newsKeyMetrics' ],
			}
		);

		expect( container ).toMatchSnapshot();

		expect( container ).toHaveTextContent( 'Insufficient permissions' );
	} );

	it( 'renders appropriate error if creating custom dimensions failed due to a generic error', () => {
		const { container } = render(
			<WidgetWithComponentProps moduleSlug="analytics-4" />,
			{
				setupRegistry: ( registry ) => {
					provideUserAuthentication( registry );
					provideUserCapabilities( registry );

					registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
						propertyID: '123456789',
						availableCustomDimensions: [],
					} );

					const error = {
						code: 'test-error-code',
						message: 'Test error message',
						data: {
							reason: 'test-error-reason',
						},
					};

					registry
						.dispatch( MODULES_ANALYTICS_4 )
						.receiveCreateCustomDimensionError(
							error,
							KEY_METRICS_WIDGETS[ KM_ANALYTICS_TOP_CATEGORIES ]
								.requiredCustomDimensions?.[ 0 ]
						);
				},
				features: [ 'newsKeyMetrics' ],
			}
		);

		expect( container ).toMatchSnapshot();

		expect( container ).toHaveTextContent( 'Analytics update failed' );
	} );

	it( 'renders report correctly if there are no errors', () => {
		const { container } = render(
			<WidgetWithComponentProps
				moduleSlug="analytics-4"
				rows={ [
					{
						field1: [ 'keyword1' ],
						field2: 0.112,
					},
					{
						field1: [ 'keyword2' ],
						field2: 0.212,
					},
					{
						field1: [ 'keyword3' ],
						field2: 0.312,
					},
				] }
				columns={ [
					{
						field: 'field1.0',
						Component: ( { fieldValue } ) => (
							<a href="http://example.com">{ fieldValue }</a>
						),
					},
					{
						field: 'field2',
						Component: ( { fieldValue } ) => (
							<strong>{ fieldValue }</strong>
						),
					},
				] }
			/>,
			{
				setupRegistry: ( registry ) => {
					provideUserAuthentication( registry );
					provideUserCapabilities( registry );

					registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
						propertyID: '123456789',
						availableCustomDimensions: [
							'googlesitekit_post_categories',
						],
					} );
				},
				features: [ 'newsKeyMetrics' ],
			}
		);

		expect( container ).toMatchSnapshot();

		expect( container ).toHaveTextContent( 'keyword1' );
	} );
} );
