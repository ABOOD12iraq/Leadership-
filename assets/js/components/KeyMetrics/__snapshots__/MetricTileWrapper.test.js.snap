// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricTileWrapper renders appropriate error if creating custom dimensions failed due to a generic error 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Top categories by pageviews
        </h2>
        <span
          class=""
          title="Categories that your site visitors viewed the most"
        >
          <svg />
        </span>
      </div>
      <h3
        class="googlesitekit-cta__title"
      >
        Analytics update failed
      </h3>
      
      
      
      
      <div
        class="googlesitekit-report-error-actions"
      >
        <button
          class="mdc-button mdc-button--raised"
          target="_self"
        >
          <span
            class="mdc-button__label"
          >
            Retry
          </span>
        </button>
        <span
          class="googlesitekit-error-retry-text"
        >
          Retry didn’t work? 
          <a
            aria-label="Learn more (opens in a new tab)"
            class="googlesitekit-cta-link googlesitekit-cta-link--external"
            href="undefined?error_id=test-error-code"
            rel="noopener noreferrer"
            target="_blank"
          >
            Learn more
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`MetricTileWrapper renders appropriate error if creating custom dimensions failed due to insufficient permissions 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Top categories by pageviews
        </h2>
        <span
          class=""
          title="Categories that your site visitors viewed the most"
        >
          <svg />
        </span>
      </div>
      <h3
        class="googlesitekit-cta__title"
      >
        Insufficient permissions
      </h3>
      
      
      
      
      <div
        class="googlesitekit-report-error-actions"
      >
        <span
          class="googlesitekit-error-retry-text"
        >
          Permissions updated? 
          <button
            class="googlesitekit-cta-link"
          >
            Retry
          </button>
        </span>
        <span
          class="googlesitekit-error-retry-text"
        >
          You’ll need to contact your administrator. 
          <a
            aria-label="Learn more (opens in a new tab)"
            class="googlesitekit-cta-link googlesitekit-cta-link--external"
            href="undefined?error_id=test-error-code"
            rel="noopener noreferrer"
            target="_blank"
          >
            Learn more
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`MetricTileWrapper renders appropriate error if required custom dimensions are not available 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Top categories by pageviews
        </h2>
        <span
          class=""
          title="Categories that your site visitors viewed the most"
        >
          <svg />
        </span>
      </div>
      <h3
        class="googlesitekit-cta__title"
      >
        No data to show
      </h3>
      
      
      
      
      <div
        class="googlesitekit-report-error-actions"
      >
        <button
          class="mdc-button mdc-button--raised"
          target="_self"
        >
          <span
            class="mdc-button__label"
          >
            Update
          </span>
        </button>
        <span
          class="googlesitekit-error-retry-text"
        >
          Update Analytics to track metric
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`MetricTileWrapper renders report correctly if there are no errors 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsTopCategories googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--table"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-km-widget-tile__title"
          >
            Top categories by pageviews
          </h3>
          <span
            class=""
            title="Categories that your site visitors viewed the most"
          >
            <svg />
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__table"
          >
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <a
                  href="http://example.com"
                >
                  keyword1
                </a>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  0.112
                </strong>
              </div>
            </div>
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <a
                  href="http://example.com"
                >
                  keyword2
                </a>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  0.212
                </strong>
              </div>
            </div>
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <a
                  href="http://example.com"
                >
                  keyword3
                </a>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  0.312
                </strong>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
